Incremental java compilation fails when renaming classname with $ character

Issue occurs since Gradle 7.3. Tested with OpenJDK 11 and 13. Tiny reproduction sample is here: https://github.com/horseunnamed/Gradle-change-detection-issue  

### Expected Behavior
Incremental java compilation is successful when classname with $ is renamed (as it is with Gradle 7.2 and earlier versions) 

### Current Behavior
After changing java class name with $ character and running incremental build we get the error:
```
> Task :compileJava FAILED
Caching disabled for task ':compileJava' because:
  Build cache is disabled
Task ':compileJava' is not up-to-date because:
  Input property 'stableSources' file .../src/main/java/issue/Class$Name.java has been removed.
  Input property 'stableSources' file .../src/main/java/issue/Main.java has changed.
  Input property 'stableSources' file .../src/main/java/issue/Class$NameChanged.java has been added.
Created classpath snapshot for incremental compilation in 0.0 secs.
Compiling with toolchain '/Library/Java/JavaVirtualMachines/openjdk-11.0.2.jdk/Contents/Home'.
Compiling with JDK Java compiler API.
.../src/main/java/issue/Main.java:5: error: cannot find symbol
        new Class$NameChanged();
            ^
  symbol:   class Class$NameChanged
  location: class Main
1 error
Incremental compilation of 3 classes completed in 0.012 secs.
:compileJava (Thread[Execution worker for ':',5,main]) completed. Took 0.02 secs.

FAILURE: Build failed with an exception.
```

### Context
On our real world Android-project we had a very harmful impact from this issue when upgrading Gradle version to 7.3.1.
Incremental annotation processor in our project generates some files with dollar sign and this issue silently led us to `ClassNotFoundException` at runtime in some cases. 
It's very serious blocker for us to upgrade from 7.2.

### Steps to Reproduce 
Please see reproduction sample: https://github.com/horseunnamed/Gradle-change-detection-issue  

1. Run build task to compile java project with classname that conatins dollar sign (e.g. `Class$Name.java`) and reference to it somewhere else (from `Main.java` in my reproduction sample)
2. Change classname and reference to it with something that still contains dollar sign
3. Rerun build task
4. Get mentioned incremental compilation error

### Your Environment
Build scan of failed incremental build from reproduction sample: https://scans.gradle.com/s/b4qlo2umn7dms

--- a/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractJavaGroovyIncrementalCompilationSupport.groovy
+++ b/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractJavaGroovyIncrementalCompilationSupport.groovy
@@ -29,7 +29,7 @@ abstract class AbstractJavaGroovyIncrementalCompilationSupport extends AbstractI
     File source(String... classBodies) {
         File out
         for (String body : classBodies) {
-            def className = (body =~ /(?s).*?(?:class|interface|enum) (\w+) .*/)[0][1]
+            def className = (body =~ /(?s).*?(?:class|interface|enum) ([\w$]+) .*/)[0][1]
             assert className: "unable to find class name"
             def f = file("src/main/${language.name}/${className}.${language.name}")
             f.createFile()

--- a/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy
+++ b/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy
@@ -492,4 +492,20 @@ sourceSets {
         then:
         failureCauseContains('Compilation failed')
     }
+
+    def "recompiles classes with \$ in class name after rename"() {
+        def firstSource = source "class Class\$Name {}"
+        source "class Main { public static void main(String[] args) { new Class\$Name(); } }"
+        outputs.snapshot { run language.compileTaskName }
+
+        when:
+        firstSource.delete()
+        source "class Class\$Name1 {}",
+            "class Main { public static void main(String[] args) { new Class\$Name1(); } }"
+        run language.compileTaskName
+
+        then:
+        outputs.deletedClasses("Class\$Name")
+        outputs.recompiledClasses("Class\$Name1", "Main")
+    }
 }

--- a/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy
+++ b/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractSourceIncrementalCompilationIntegrationTest.groovy
@@ -493,6 +493,7 @@ sourceSets {
         failureCauseContains('Compilation failed')
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/19257")
     def "recompiles classes with \$ in class name after rename"() {
         def firstSource = source "class Class\$Name {}"
         source "class Main { public static void main(String[] args) { new Class\$Name(); } }"

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java
@@ -56,9 +56,20 @@ public class FileNameDerivingClassNameConverter implements SourceFileClassNameCo
             return sourcePaths;
         }
 
+        Set<String> paths = fileExtensions.stream()
+            .map(fileExtension -> className.replace('.', '/') + fileExtension)
+            .collect(Collectors.toSet());
+
+        // Classes with $ may be inner classes
         int innerClassIdx = className.indexOf("$");
-        String baseName = innerClassIdx > 0 ? className.substring(0, innerClassIdx) : className;
-        return fileExtensions.stream().map(fileExtension -> baseName.replace('.', '/') + fileExtension).collect(Collectors.toSet());
+        if (innerClassIdx > 0) {
+            String baseName = className.substring(0, innerClassIdx);
+            paths.addAll(fileExtensions.stream()
+                .map(fileExtension -> baseName.replace('.', '/') + fileExtension)
+                .collect(Collectors.toSet()));
+        }
+
+        return paths;
     }
 
 }

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java
@@ -18,6 +18,7 @@ package org.gradle.api.internal.tasks.compile.incremental.recomp;
 import org.apache.commons.lang.StringUtils;
 
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Collectors;
 
@@ -57,19 +58,22 @@ public class FileNameDerivingClassNameConverter implements SourceFileClassNameCo
         }
 
         Set<String> paths = fileExtensions.stream()
-            .map(fileExtension -> className.replace('.', '/') + fileExtension)
+            .map(fileExtension -> classNameToRelativePath(className, fileExtension))
             .collect(Collectors.toSet());
 
         // Classes with $ may be inner classes
         int innerClassIdx = className.indexOf("$");
         if (innerClassIdx > 0) {
             String baseName = className.substring(0, innerClassIdx);
-            paths.addAll(fileExtensions.stream()
-                .map(fileExtension -> baseName.replace('.', '/') + fileExtension)
-                .collect(Collectors.toSet()));
+            fileExtensions.stream()
+                .map(fileExtension -> classNameToRelativePath(baseName, fileExtension))
+                .forEach(paths::add);
         }
 
         return paths;
     }
 
+    private String classNameToRelativePath(String className, String fileExtension) {
+        return className.replace('.', '/') + fileExtension;
+    }
 }

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/FileNameDerivingClassNameConverter.java
@@ -18,7 +18,6 @@ package org.gradle.api.internal.tasks.compile.incremental.recomp;
 import org.apache.commons.lang.StringUtils;
 
 import java.util.Collections;
-import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Collectors;
 

